costmap_2d/plugins/obstacle_layer.cpp

```diff
@@ -531,6 +538,54 @@ void ObstacleLayer::raytraceFreespace(const Observation& clearing_observation, d
     double a = wx - ox;
     double b = wy - oy;
 
+    // calculate raytrace starting point
+    double ray_length = sqrt(a * a + b * b);
+
+    // skip if the ray length is even shorter than min_raytrace_range
+    if(ray_length < min_raytrace_dist)
+    {
+      continue;
+    }
+
+    double raytrace_x = 0.0, raytrace_y = 0.0;
+    raytrace_x = ox + min_raytrace_dist * a / ray_length;
+    raytrace_y = oy + min_raytrace_dist * b / ray_length;
+
+    // check if the raytrace starting point is within the map boundary
+    if (raytrace_x < origin_x)
+    {
+      double t = (origin_x - ox) / a;
+      raytrace_x = origin_x;
+      raytrace_y = oy + b * t;
+    }
+
+    if (raytrace_y < origin_y)
+    {
+      double t = (origin_y - oy) / b;
+      raytrace_x = ox + a * t;
+      raytrace_y = origin_y;
+    }
+
+    if (raytrace_x > map_end_x)
+    {
+      double t = (map_end_x - ox) / a;
```

Comment:
Is there a way to share the code between these two?  It's ok for now, just seems error prone.  At least let's put in a comment here to change the other file if a bug is found.  Best fix is to isolate this to another shared file.